M2/Macaulay2/packages/HomologicalShiftIdeals.m2

```diff
@@ -0,0 +1,1001 @@
+        J:=f I;
+        Res:=resolution J;
+        d:=degrees Res_r;
+        M:={};
+        for w in d do (
+              M=append(M,toMonomial(S,w));
+              M;
+        );
+        unique M
+    );
+
+-------------------------------------------------------------------------------------------
+--  Compute the kth homological shift ideal of a monomial ideal
+-------------------------------------------------------------------------------------------
+mydoc = concatenate(mydoc,///
+Node
+  Key
+    HS
+    (HS,Ideal,ZZ)
+  Headline
+    compute the kth homological shift ideal of a monomial ideal
+  Usage
+    HS(I,k)
+  Inputs
+    I:Ideal
+      a monomial ideal $I\subset S=K[x_1,\dots,x_n]$, $K$ a field
+    k:ZZ
+      an integer $k$
+  Outputs
+    l:List
+      the $k$th homological shift ideal of $I$
+  Description
+    Text
+      Given a monomial ideal $I$ of a polynomial ring $S=K[x_1,\ldots,x_n]$ with minimal multigraded free resolution $$\mathbb{F}:\cdots\rightarrow\bigoplus_{j=1}^{\beta_k(I)}S(-\textbf{a}_{k,j})\rightarrow\bigoplus_{j=1}^{\beta_{k-1}(I)}S(-\textbf{a}_{k-1,j})\rightarrow\cdots\rightarrow\bigoplus_{j=1}^{\beta_0(I)}S(-\textbf{a}_{0,j})\rightarrow I\rightarrow 0,$$ the $k\text{th}$ homolofical shift ideal of $I$ is $$\text{HS}_k(I)=(\textbf{x}^{\textbf{a}_{k,j}}:j=1,\dots,\beta_k(I)).$$
+    Example
+      S = QQ[x_1..x_5];
+      I = ideal(x_1,x_2,x_3,x_4)*ideal(x_3,x_4,x_5);
+      HS(I,0)
+      HS(I,1)
+      HS(I,2)
+      HS(I,3)
+      HS(I,4)
+      HS(I,5)
+///);
+
+HS = method(TypicalValue=>Ideal)
+HS (Ideal,ZZ) := (I,k) -> (
+        if not isMonomialIdeal(I) then error
+        else S:=ring I;
+        M:=multigradedShifts(I,k);
+        H:=trim ideal(M);
+        if H==(0) or H==(1) then return ideal(0_S)
+        else return H
+    );
+
+-------------------------------------------------------------------------------------------
+--  Compute the socle elements of a monomial ideal
+-------------------------------------------------------------------------------------------
+mydoc = concatenate(mydoc,///
+Node
+  Key
+    socle
+    (socle,Ideal)
+  Headline
+    compute the socle of a monomial ideal
+  Usage
+    socle(I)
+  Inputs
+    I:Ideal
+      a monomial ideal $I\subset S=K[x_1,\dots,x_n]$, $K$ a field
+  Outputs
+    l:List
+      the socle elements of $I$
+  Description
+    Text
+      Given a monomial ideal $I$ of a polynomial ring $S=K[x_1,\ldots,x_n]$, the socle set of $I$ is the set given by all monomials $v$ such that $v\notin I$ and $x_iv\in I$ for all $i=1,\dots,n$.
+    Example
+      S = QQ[x_1..x_5];
+      I = ideal(x_1,x_2,x_3,x_4)*ideal(x_3,x_4,x_5);
+      socle I
+///);
+
+socle = method(TypicalValue=>List)
+socle (Ideal) := I -> (
+        if not isMonomialIdeal(I) then error
+        else S:=ring I;
+        n:=numgens S-1;
+        soc:=multigradedShifts(I,n);
+        L:={};
+        if isFullySupported(I)==false then return "expected a fully supported monomial ideal"
```

Comment:
This could be changed to `if not isFullySupported I then error "expected a fully supported monomial ideal';`